test/bench_cache.lua

```diff
@@ -0,0 +1,180 @@
+    end
+end
+
+local sharding_func_body = [[
+function(shard_key)
+    local digest = require('digest')
+    if type(shard_key) ~= 'table' then
+        return digest.crc32(tostring(shard_key))
+    else
+        local crc32 = digest.crc32.new()
+        for _, v in ipairs(shard_key) do
+            crc32:update(tostring(v))
+        end
+        return crc32:result()
+    end
+end
+]]
+
+local primary_index = {
+    type = 'HASH',
+    unique = true,
+    parts = {
+        {path = 'string_nonnull', is_nullable = false, type = 'string'},
+        {path = 'unsigned_nonnull', is_nullable = false, type = 'unsigned'},
+    },
+    name = 'primary'
+}
+
+local bucket_id_idx = {
+    type = 'TREE',
+    unique = false,
+    parts = {{path = 'bucket_id', type = 'unsigned', is_nullable = false}},
+    name = 'bucket_id'
+}
+
+local function space_init()
+    db.drop_all()
+
+    local space = {
+        engine = 'memtx',
+        is_local = true,
+        temporary = false,
+        format = table.deepcopy(helper.test_space_format())
+    }
+    table.insert(space.format, 1, {
+        name = 'bucket_id', type = 'unsigned', is_nullable = false
+    })
+
+    space.indexes = {
+        table.deepcopy(primary_index),
+        table.deepcopy(bucket_id_idx)
+    }
+    space.sharding_key = {'unsigned_nonnull', 'integer_nonnull'}
+    local schema = {
+        spaces = {
+            space = space,
+        }
+    }
+
+    return schema
+end
+
+local function run_body()
+    db.init()
+    local schema = space_init()
+    schema.spaces.space.sharding_func = {
+        body = sharding_func_body
+    }
+
+    local _, err = ddl.set_schema(schema)
+    if err then
+        print(err)
+        os.exit()
+    end
+
+    for i=1,ITERS do
+        local _, err = ddl.bucket_id('space', i)
+        if err then
+            print(err)
+            os.exit()
+        end
+    end
+
+    print("Done")
+    os.exit()
+end
+
+local function run_baseline()
+    db.init()
+    local schema = space_init()
+    schema.spaces.space.sharding_func = {
+        body = sharding_func_body
+    }
+
+    local _, err = ddl.set_schema(schema)
+    if err then
+        print(err)
+        os.exit()
+    end
+
+    for i=1,ITERS do
+        mpcrc32(i)
+        if err then
+            print(err)
+            os.exit()
+        end
+    end
+
+    print("Done")
+    os.exit()
+end
+
+local function run_name()
+    db.init()
+    local schema = space_init()
+    local sharding_func_name = 'mpcrc32'
+    rawset(_G, sharding_func_name, mpcrc32)
+    schema.spaces.space.sharding_func = sharding_func_name
+
+    local _, err = ddl.set_schema(schema)
+    if err then
+        print(err)
+        os.exit()
+    end
+
+    for i=1,ITERS do
+        local _, err = ddl.bucket_id('space', i)
+        if err then
+            print(err)
+            os.exit()
+        end
+    end
+
+    print("Done")
+    os.exit()
+end
+
+local function main()
+    local getopt = require('posix.unistd').getopt
```

Comment:
Is it from luaposix? It is not something that we have installed by default. I would prefer to don't depend on it.

We can either look over `arg` ourself (here it is quite simple) or use `internal.argparse` (which is internal, sure, but it changes rarely and built-in).

Reply:
Ok, I will rewrite with manually args parsing. But I prefer using getopt() where possible.

Reply:
fixed